### PR TITLE
Jason/issue 327

### PR DESCRIPTION
**Pull request by duidae**
fix [jira #327](https://twreporter-org.atlassian.net/browse/TWREPORTER-327)

need companion PR in `plate` repo

P.S.
`RelationshipFieldForCategory.js`這個檔案是直接copy `RelationshipField.js`，目的是為了新舊相容，因為原分類是使用RelationshipField這個物件，但同時RelationshipField也被好幾個欄位共用，這次分類新舊相容需額外判斷的部分就加在`RelationshipFieldForCategory.js`裡(可以diff RelationshipFieldForCategory.js/RelationshipField.js)就不影響其他欄位，migration完直接刪掉`RelationshipFieldForCategory.js`即可

### CODE REVIEW

File: fields/types/categorySet/CategorySet.js [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        return;
+      }
+      let categoryOptions = [];
+      let subcategoryOptionsMap = new Map([]);
+      async.each(data.results, (category, callback) => {
+        if (category && category.id && category.name && category.fields && Array.isArray(category.fields.subcategory) && category.fields.subcategory.length > 0) {
+          categoryOptions.push({ value: category.id, label: category.name });
+          const subcategoryIDs = category.fields.subcategory;
+          getSubcategoryOptions(subcategoryIDs, (err, results) => {
+            if (err) {
```
Comment (Aylie-Chou) [inline fields/types/categorySet/CategorySet.js:85]: Question: Would we consider to retry `getSubcategoryOptions` function or hint user to reload if `getSubcategoryOptions` fail ?
On the side of user, category set cannot be done without subcategory options.

File: fields/types/categorySet/CategorySetType.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * @api public
+ */
+
+categorySet.prototype.updateItem = function(item, data, callback) {
+  var value = this.getValueFromData(data);
+  if (value && value !== '') {
+    var dObj = {};
+    try {
+      dObj = JSON.parse(value);
```
Comment (Aylie-Chou) [inline fields/types/categorySet/CategorySetType.js:37]: Could we add some check with category set item in dObj to filter out `{ category: undefined, subcategory: undefined }` case ?

**Review by Aylie-Chou — CHANGES_REQUESTED** [review]
Almost there ! Just some error handles to be done throughly.
Thanks for the troubble shooting, it really saves my day 🙏

File: fields/types/categorySet/CategorySetType.js (hunk 2 above)
Comment (duidae) [reply fields/types/categorySet/CategorySetType.js:37]: I filter that ahead in the `<input>` component via `JSON.stringify(value.filter(categorySet => categorySet && categorySet.category))`
https://github.com/twreporter/keystone/blob/4f6a9726546c3f5517856ec16f5370280950f2d8/fields/types/categorySet/CategorySet.js#L164

but fine to reinforce again in https://github.com/twreporter/keystone/pull/220/commits/b42d8ecaa70568b48f29e4add9a60bc87441db31

File: fields/types/categorySet/CategorySet.js (hunk 1 above)
Comment (duidae) [reply fields/types/categorySet/CategorySet.js:86]: alert added in https://github.com/twreporter/keystone/pull/220/commits/4f6a9726546c3f5517856ec16f5370280950f2d8

**Comment by duidae** [comment]
> Almost there ! Just some error handles to be done throughly. Thanks for the troubble shooting, it really saves my day 🙏

no problem! comments are fixed, please have a look, thanks.

**Review by Aylie-Chou — APPROVED** [review]
LGTM 👍

**Comment by duidae** [comment]
@Aylie-Chou append fix for spec change in https://github.com/twreporter/keystone/pull/220/commits/990f6611ec51584ee132c0f68a224c5f46a9bc2b, please help review, thanks!

**Comment by Aylie-Chou** [comment]
LGTM !

**Pull request merged**